intermediate_source/TP_tutorial.rst

```diff
@@ -0,0 +1,282 @@
+* Run sharded computation for the parallelized layers to save compute/memory. (i.e. ``nn.Linear``, ``nn.Embedding``)
+
+
+When and Why you should apply Tensor Parallel
+------------
+The PyTorch Fully Sharded Data Parallel (FSDP) could scale the model training to certain number of GPUs already, but when further scaling the model with regard to model size and number of GPUs, there are many more challenges need to be resolved, where it might need Tensor Parallelism (TP) to combine with FSDP:
+The world size (# GPUs) becoming too large (> 128/256 GPUs) and the FSDP collectives (i.e. ``allgather``) are being dominated by ring latency. TP/SP on top of FSDP would cut the FSDP world size by 8 thereby reducing the latency costs by 8.
+Hit data parallelism limit where you can not raise the global batch size to be above the # GPUs due to both convergence and GPU memory limitations, Tensor/Sequence Parallelism is the only known way to “ballpark” the global batch size and continue scaling with more GPUs. This means both model size and number of GPUs could continue to scale.
+For certain types of models, when local batch size becomes smaller TP/SP gives more flop friendly matrix multiplication shapes.
+
+So when pre-training, how easy is it to hit those limits? Today to pre-train a Large Language Model (LLM) with billions/trillions of tokens it could take months even with thousands of GPUs.
+It will always hit limitation 1 when training LLM on a large scale. For example, Llama 2 70B trained with 2k GPUs for 35 days, multi-dimensional parallelisms are needed at 2k scale.
+When the Transformer model becomes larger (i.e. Llama2 70B), it will also quickly hit the limitation 2: One could not use FSDP alone with even local batch_size=1 due to memory and convergence constraints. For example, Llama 2 global batch size is 1k, so data parallelism alone can not be used at 2k GPUs.
+
+
+How to apply Tensor Parallel
+---------------
+
+
+To demonstrate how to apply tensor parallelism using the PyTorch native Tensor Parallel APIs, let us look at a common Transformer model. In this tutorial, we use the most recent `Llama2 model <https://github.com/pytorch/examples/blob/main/distributed/tensor_parallelism/>__` as a reference Transformer model implementation, as it is also widely used in the community.
+
+Since Tensor Parallel shards individual tensors over a set of devices, we would need to set up the distributed environment (i.e NCCL communicators) first. Tensor Parallelism is a SPMD sharding algorithm similar to PyTorch DDP/FSDP, and it under the hood leverages the PyTorch DTensor to perform sharding. It also utilizes the DeviceMesh abstraction (which under the hood manages ProcessGroups) for device management and sharding. To see how to utilize DeviceMesh to set up multi-dimensional parallelisms, please refer to this tutorial. Tensor Parallel usually works within each host, so let us first initialize a DeviceMesh that connects 8 GPUs within a host.
```

Comment:
```suggestion
Since Tensor Parallel shards individual tensors over a set of devices, we would need to set up the distributed environment (such as NCCL communicators) first. Tensor Parallelism is a Single-Program Multiple-Data (SPMD) sharding algorithm similar to PyTorch DDP/FSDP, and it under the hood leverages the PyTorch DTensor to perform sharding. It also utilizes the DeviceMesh abstraction (which under the hood manages ProcessGroups) for device management and sharding. To see how to utilize DeviceMesh to set up multi-dimensional parallelisms, please refer to `this tutorial <https://pytorch.org/tutorials/recipes/distributed_device_mesh.html>`__. Tensor Parallel usually works within each host, so let us first initialize a DeviceMesh that connects 8 GPUs within a host.
```